python/python/lance/cache.py

```diff
@@ -0,0 +1,63 @@
+#  Copyright (c) 2023. Lance Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Iterable, Optional, Union
+
+import pyarrow as pa
+
+
+class CachedDataset:
+    def __init__(
+        self,
+        stream: Iterable[pa.RecordBatch],
+        cache: Optional[Union[str, Path, bool]] = None,
+    ):
+        self.cache_dir: Optional[TemporaryDirectory] = None
+        if cache is None or cache is True:
+            self.cache_dir = TemporaryDirectory(
+                prefix="lance-torch-dataset",
+            )
+        elif isinstance(cache, (str, Path)):
+            self.cache_dir = TemporaryDirectory(
+                prefix="lance-torch-dataset",
+                dir=cache,
+            )
+        else:
+            raise ValueError(f"Unsupported cache type: {type(cache)}")
+        self.cache_file = None
+        self.stream = stream
+
+    def __del__(self):
+        if self.cache_dir is not None:
+            self.cache_dir.cleanup()
+            self.cache_dir = None
+
+    def __iter__(self):
```

Comment:
nit: do we wannt check and make sure the second stream doesn't iter faster than the first stream?

Reply:
Ok, added one.